### PR TITLE
BUGFIX - AttributeError raised in StataReader.value_labels()

### LINKED ISSUE #19417: value_labels fails to assign value_label_dict, leads to AttributeError
#### Code Sample, a copy-pastable example if possible

```python
def test_StataReader(stata_path):
    """
    Args:
        stata_path - path to literally any valid stata file.
    """
    # Step 1: Delay reading value labels, which is done automaticaly when
    # decoding categoricals. Otherwise this only works if we have a stata file
    # with format verison <= 108
    reader = pandas.io.stata.StataReader(stata_path,
                                         convert_categoricals=False)
    # Patch format to be <= 108
    # https://github.com/pandas-dev/pandas/blob/master/pandas/io/stata.py#L1341-L1344
    with patch.object(reader, 'format_version', 108):
        # Trigger AttributeError when _read_value_labels immediately returns without
        # assigning value_label_dict, which value_labels() expects to exist
        # https://github.com/pandas-dev/pandas/blob/master/pandas/io/stata.py#L1745
        reader.value_labels()
```

#### Problem description

An unlikely code path will lead to an AttributeError.

#### Expected Output

I thin it is reasonable to expect `value_labels()` to return either `None` or `{}` when the stata file's `format_version` is <= 108

#### Output of ``pd.show_versions()``

<details>

>>> pandas.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.1.final.0
python-bits: 64
OS: Darwin
OS-release: 17.3.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8
LOCALE: en_US.UTF-8

pandas: 0.21.0
pytest: 3.0.7
pip: 9.0.1
setuptools: 27.2.0
Cython: None
numpy: 1.13.3
scipy: None
pyarrow: None
xarray: None
IPython: None
sphinx: None
patsy: None
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: 1.0.0
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: None
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by miker985**
- [x] closes #19417 
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Returns empty dict instead of raising AttributeError for very old Stata files. Includes test against very old Stata file.

### CODE REVIEW

File: doc/source/whatsnew/v0.23.0.txt [hunk 1]
```diff
@@ -530,6 +530,8 @@ I/O
 - Bug in :func:`DataFrame.to_parquet` where an exception was raised if the write destination is S3 (:issue:`19134`)
 - :class:`Interval` now supported in :func:`DataFrame.to_excel` for all Excel file types (:issue:`19242`)
 - :class:`Timedelta` now supported in :func:`DataFrame.to_excel` for xls file type (:issue:`19242`, :issue:`9155`)
+- Bug in :class:`pandas.io.stata.StataReader` raising ``AttributeError`` when ``value_labels()`` called with very old files. Now returns an empty dict (:issue:`19417`)
```
Comment (TomAugspurger) [inline doc/source/whatsnew/v0.23.0.txt]: We don't have an entry for `StataReader` in the API docs, but we do have `StataReader.value_labels`. So could you reword this as

    - Bug in :meth:`pandas.io.stata.StataReader.value_labels` raising an ``AttributeError`` when called on very old files. Now returns an empty dict (:issue:`19417`)

File: pandas/tests/io/test_stata.py [hunk 2]
```diff
@@ -589,6 +589,15 @@ def test_105(self):
         df0['psch_dis'] = df0["psch_dis"].astype(np.float32)
         tm.assert_frame_equal(df.head(3), df0)
 
+    def test_value_labels_old_format(self):
+        # GH 19417
+        #
+        # Test that value_labels() returns an empty dict if the file format
+        # predates supporting value labels.
+        dpath = os.path.join(self.dirpath, 'S4_EDUC1.dta')
+        reader = StataReader(dpath)
```
Comment (jreback) [inline pandas/tests/io/test_stata.py]: need to close the reader (or do this in a context manager) see other tests

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/io/stata.py [hunk 3]
```diff
@@ -1341,12 +1341,14 @@ def _null_terminate(self, s):
...
-        if self.format_version <= 108:
-            # Value labels are not supported in version 108 and earlier.
-            return
         if self._value_labels_read:
             # Don't read twice
             return
+        if self.format_version <= 108:
+            # Value labels are not supported in version 108 and earlier.
+            self._value_labels_read = True
+            self.value_label_dict = dict()
```
Comment (bashtage) [inline pandas/io/stata.py]: Was there a preference stated recently to use `{}` rather than `dict()`?

Comment (jreback) [reply pandas/io/stata.py]: yeah can use ``{}``

Comment (miker985) [reply pandas/io/stata.py]: `dict()` is internally consistent with the function.

File: pandas/tests/io/test_stata.py (hunk 2 above)
Comment (miker985) [reply pandas/tests/io/test_stata.py]: `close()`d

File: doc/source/whatsnew/v0.23.0.txt (hunk 1 above)
Comment (miker985) [reply doc/source/whatsnew/v0.23.0.txt]: Changed

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. rebasing, ping when green.

**Comment by jreback** [comment]
thanks!
